Add support for building an ld.so.cache from a filesystem

dannf opened the pull request.

dannf committed:
Bump github.com/stretchr/testify from v1.8.1 to v1.10.0

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Make LDSOCacheFile Write() use an io.Writer

This will make it easier to use the fs.FS interface

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Add support for generating a cache file from dirs in a fs.FS

This is currently full of unfilterable debug statements.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Include .so files with no SONAME

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Add support for parsing ld.so.conf files

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Fix off-by-one error

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
If filesystem implements ReaderAt, use it

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Add commented-out test that generates a cache file for the host system

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
staticcheck: this value of pos is never used (SA4006)

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
staticcheck: error strings should not be capitalized (ST1005)

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
 	fmt.Printf("  Size [%d]\n", shdr.Size)
 }
 
+func parse_lib_filename(realname string) (string, string, error) {
```
https://google.github.io/styleguide/go/guide#mixedcaps

dannf replied:
fixed, thanks!

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+	return name, ver, nil
+}
+
+func AddLDSOCacheEntriesForDir(fsys fs.FS, libdir string, entries *[]LDSOCacheEntry) (error) {
```
Should have a comment: https://go.dev/wiki/CodeReviewComments#doc-comments

I don't think you need `entries` to be a pointer to use it the way you're using it. I'd also probably have this return a slice rather than updating `entries` across multiple calls (or just inline this implementation into `AddLDSOCacheEntriesForDirs` and have that return a slice).

You don't need `(error)` as a return type, just `error`.

dannf replied:
> Should have a comment: https://go.dev/wiki/CodeReviewComments#doc-comments

Added

> I don't think you need `entries` to be a pointer to use it the way you're using it.

I really need to read about Go calling conventions. But I couldn't figure out a way to do that. If I just naively drop the pointer syntax, it runs, but nothing gets added to the list.

> I'd also probably have this return a slice rather than updating `entries` across multiple calls

Yeah, that's how I originally wrote it. But when I compared the output to `ldconfig`'s, I realized that `ldconfig` avoids adding libraries that area already there, so I do need the full context.

>  (or just inline this implementation into `AddLDSOCacheEntriesForDirs` and have that return a slice).

yes, returning a slice is better, thanks! I chose not to inline the ForDir into the ForDirs because I feel that function is already too long and nested. In fact, I want to go further and add unroll into `AddLDSOCacheEntryForLib` and `AddLDSOCacheEntryForSoname`, which should get it closer to editor-window-size.

> You don't need `(error)` as a return type, just `error`.

fixed!

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+
+func AddLDSOCacheEntriesForDir(fsys fs.FS, libdir string, entries *[]LDSOCacheEntry) (error) {
+	// fs.FS wants all file paths to be relative
+	libdir = strings.TrimLeft(libdir, "/")
```
You probably want `strings.TrimPrefix`.

dannf replied:
My intent was to trim all leading `/`, so I could handle the case of someone passing in `///usr/lib`. But it'd be more clear if I was doing this with `filepath` logic. So: https://github.com/chainguard-dev/ldso-cache/commit/9a16ab2bda1580de7d93dd07589666d06721cbf4

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+	libdir = strings.TrimLeft(libdir, "/")
+	dirents, err := fs.ReadDir(fsys, libdir)
+	if err != nil {
+		// It is OK for a directory to not exist
```
If the only acceptable/ignorabe error is that it doesn't exist, you should check for that error with `errors.Is(err, fs.ErrNotExist)`

dannf replied:
nice! fixed.

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+		realname := dirent.Name()
+		fullpath := filepath.Join(libdir, realname)
+		mode := dirent.Type()
+		is_link := (mode & fs.ModeSymlink != 0)
```
https://google.github.io/styleguide/go/guide#mixedcaps

dannf replied:
fixed

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+		}
+		libf, err := fsys.Open(fullpath)
+		if err != nil {
+			continue
```
Is it okay to ignore this error?

dannf replied:
I checked to see what `ldconfig` does by `chmod 000`'ing a lib:

```
$ ldconfig -C /tmp/cache
/sbin/ldconfig.real: Input file /lib/x86_64-linux-gnu/libpagemaker-0.0.so.0.0.4 not found.
$ echo $?
0
```

So I've added a warning here too.

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+			libfReaderAt = bytes.NewReader(buf)
+		}
+		elflibf, err := elf.NewFile(libfReaderAt)
+
```
unnecessary newline

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+		flags |= FLAG_ELF_LIBC6
+		sonames, err := elflibf.DynString(elf.DT_SONAME)
+		if err != nil {
+			continue
```
Is it okay to ignore this error?

dannf replied:
I think so, because `ldconfig` does:

```
$ sudo touch /usr/lib/x86_64-linux-gnu/libspider.so.9
$ ldconfig -C /tmp/cache 
/sbin/ldconfig.real: File /lib/x86_64-linux-gnu/libspider.so.9 is empty, not checked.
$ echo notanelf | sudo tee /usr/lib/x86_64-linux-gnu/libspider.so.9
notanelf
$ ldconfig -C /tmp/cache 
$ echo $?
0
$ ldconfig -p -N -C /tmp/cache | grep spider
$
```

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+			fmt.Printf("DEBUG: %s has no DT_SONAME, using %s as an SONAME\n", realname, realname)
+		}
+
+		for _, soname := range(sonames) {
```
`range` isn't a function, you want: `range sonames`

dannf replied:
25 years of python is hard to break! fixed

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+				continue
+			}
+			dupe_found := false
+			for _, existing := range *entries {
```
This is quadratic and will be slow if `entries` is huge. Probably not something that will happen that often, but I'd use a map to track entry names so that it's faster to check dupes.

dannf replied:
done

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -337,3 +545,69 @@ func (hdr *LDSORawCacheHeader) Write(w io.Writer) error {
+	var libpaths []string
+	var seenpaths []string
+
+	lines := strings.Split(string(contents), "\n")
```
Use [`bufio.Scanner`](https://pkg.go.dev/bufio#Scanner) for this to avoid buffering.

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -337,3 +545,69 @@ func (hdr *LDSORawCacheHeader) Write(w io.Writer) error {
+		if err != nil {
+			return nil, err
+		}
+		if slices.Contains(seenpaths, realpath) {
```
I'd use a map for this too.

dannf replied:
I thought I agreed, until I implemented it:
https://github.com/dannf/ldso-cache/commit/20acf942c31d8e2f1a1abec4d3b483c7a9bae574

Since this function is recursive, it needs to return a map if it uses one. And that means I need a wrapper function around it to give the caller back a slice. And what we're actually storing here is a list of directories that may contain libraries - which should almost always be in the single digits.

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -337,3 +545,69 @@ func (hdr *LDSORawCacheHeader) Write(w io.Writer) error {
+
+		if err != nil {
+			return nil, err
+		}
```
I think this is accidentally here?

dannf replied:
I realized I should kill most of this section:
https://github.com/chainguard-dev/ldso-cache/commit/ac6bcbc05669da49d0ac6f4cef591341b694996c

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+		ver = pieces[len(pieces)-1]
+	}
+
+	fmt.Printf("RETURNING %s for %s\n", name, realname)
```
There's a `log` package I'd use for warnings and stuff which prints to stderr. `fmt.Printf` goes to stdout.

dannf replied:
using `log` now

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+			// Ugly: Work around lack of ReaderAt support by
+			// reading the entire file into memory
+			var buf []byte
+			buf, err = fs.ReadFile(fsys, fullpath)
```
You don't need the `var` here, this can be `buf, err := ...`

dannf replied:
fixed!

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+	if err != nil {
+		return nil, err
+	}
+	var magic [17]byte
+	var version [3]byte
+	copy(magic[:], ldsoMagic)
+	copy(version[:], ldsoVersion)
```
This is a little strange, you could inline this below as:

```
		Magic:   [17]byte([]byte(ldsoMagic)),
		Version: [3]byte([]byte(ldsoVersion)),
```

dannf replied:
done!

jonjohnsonjr commented on ldsocache.go:
```diff
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
+
+	return &cf, nil
+}
+
```
I can't comment on LoadCacheFile because it's below the diff but it should take an `io.ReadSeeker` or `io.ReaderAt` instead of a file path.

dannf replied:
fixed

dannf committed:
Rename parse_lib_filename() to ParseLibFilename()

Observe MixedCaps naming scheme.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Use filepath logic for path instead of string tinkering

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: error if ReadDir fails for other reasons

It's OK if a directory doesn't exist. But if it exists, and can't
be read for some other reason (permissions?), then we should error.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Fix range usage

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
ParseLDSOConf: Don't bother checking if library paths exist

We're just parsing the config, we're not its mother.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: Remove unnecessary var declaration

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
BuildCacheFileForDirs: Simplify population of byte fields

Requires bumping go1.20:

cannot convert []byte(ldsoMagic) (value of type []byte) to type [17]byte: conversion of slice to array requires go1.20 or later (-lang was set to go1.18; check go.mod)

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: Remove parens around single return

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDirs: Remove parens around single return

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
add comment blocks for non-trivial functions

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDirs: Create the return slice

Don't require the caller to provide one.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
use mixedcaps instead of underscores in variable names

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: emit warning when lib can't be opened

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: remove unnecessary newline

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDir: use a map instead of a slice

For faster lookups.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
ParseLibFilename: Add tests, and make them work

Added tests that uncovered bugs in the implementation.
Fixed those bugs.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Use log package

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
Make debug messages optional

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
ldso-cache: fix alignment math

Our alignment formula would sometimes go backwards:

>>> pos = 3391
>>> (pos & -16) + 8
3384

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
AddLDSOCacheEntriesForDirs: Sort libraries for reproducibility

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
ldso-cache: Use CamelCaps

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
extractShlibName: drop error return, which is always nil

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
golangci-lint: Consider pre-allocating `libpaths` (prealloc)

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
go fmt

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

dannf committed:
LoadCacheFile(): use a ReadSeeker instead of a file path

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>
